File: UPCH_SAMVGW_INTELGRV_LNUX_02/LittleEndian/IEX/IEX_SRC_FILES/IEX.ElementaryActions.FunctionalityCS/EAs/ManualRecording/ModifyManualRecording.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using FailuresHandler;
using IEX.ElementaryActions.Functionality;
using System.Globalization;
using IEX.ElementaryActions.EPG;

namespace EAImplementation
{
    /// <summary>
    ///   Modify Manual Recording From Planner
    /// </summary>
    public class ModifyManualRecording : IEX.ElementaryActions.BaseCommand
    {

        private IEX.ElementaryActions.EPG.SF.UI EPG;

        private IEX.ElementaryActions.Functionality.Manager _manager;
        private string _EventKeyName;
        private string _StartTime;
        private string _EndTime;
        private string _ChannelName;
        private int _Days;
        private Boolean _IsFirstTime;
		private bool recordNameChangeSupported = true;

        private EnumFrequency _Frequency;

        /// <param name="EventKeyName">Key Of The Event</param>
        /// <param name="StartTime">Channel Name</param>
        /// <param name="EndTime">Channel Number If Entered Doing DCA</param>
        /// <param name="ChannelName">Optional Parameter Default = "" : The Channel Name</param>
        /// <param name="Days">Optional Parameter Default = 0 : Adds Days From Current Date</param>
        /// <param name="Frequency">Optional Parameter Default = ONE_TIME</param>
        /// <param name="pManager">Manager</param>
        /// <remarks>
        /// Possible Error Codes:
        /// <para>300 - NavigationFailure</para> 
        /// <para>301 - DictionaryFailure</para> 
        /// <para>302 - EmptyEpgInfoFailure</para> 
        /// <para>304 - IRVerificationFailure</para> 
        /// <para>305 - PCATFailure</para> 
        /// <para>309 - GetEpgTimeFailure</para>    
        /// <para>310 - GetEpgDateFailure</para>    
        /// <para>322 - VerificationFailure</para> 
        /// <para>328 - INIFailure</para>  
        /// <para>330 - TelnetFailure</para> 
        /// <para>331 - CopyFileFailure</para> 
        /// <para>332 - NoValidParameters</para> 
        /// <para>334 - VideoNotPresent</para>  
        /// <para>339 - RecordEventFailure</para>  
        /// <para>349 - ReturnToLiveFailure</para> 
        /// <para>350 - ParsingFailure</para> 
        /// </remarks>
        public ModifyManualRecording(string EventKeyName, string StartTime, string EndTime, string ChannelName, int Days, EnumFrequency Frequency, Boolean IsFirstTime, IEX.ElementaryActions.Functionality.Manager pManager)
        {
            this._EventKeyName = EventKeyName;
            this._StartTime = StartTime;
            this._EndTime = EndTime;
            this._ChannelName = ChannelName;
            this._Days = Days;
            this._Frequency = Frequency;
            this._IsFirstTime = IsFirstTime;
            this._manager = pManager;
            EPG = this._manager.UI;
        }

        /// <summary>
        ///  EA Execution
        /// </summary>
        protected override void Execute()
        {

            IEXGateway._IEXResult res;
            string EventName = "";
            EpgEvent eventToBeModified = new EpgEvent(EPG.Utils);
            String EPGDate = "";
            String EPGEventdate = "";
            String PCATEventDate = "";
            String EPGDateFormat = "";
            String PCATDateFormat = "";
            String milestoneDateFormat = "";
            bool Current = true;
            string EventSource = "ManualFuture";
            string Frequency = "";
            int Occurences = 0;
            long newDuration = 0;
            String tuneToService = "";
            string TodayDateInEPGFormat = "";

            tuneToService = EPG.Utils.GetValueFromProject("MANUAL_RECORDING", "TUNE_TO_SERVICE_BEFORE_RECORD");
            milestoneDateFormat = EPG.Utils.GetValueFromProject("EPG", "MILESTONE_DATE_FORMAT");
			try
            {
                string recordNameChangeSupportedVal = EPG.Utils.GetValueFromProject("MANUAL_RECORDING", "RECORD_NAME_CHANGE_SUPPORTED");
                if (recordNameChangeSupportedVal != null)
                {
                    recordNameChangeSupported = Convert.ToBoolean(recordNameChangeSupportedVal);
                }
            }
            catch
            {
                recordNameChangeSupported = true;
            }
            PCATDateFormat = EPG.Utils.GetDateFormatForEventDictionary();
            EPGDateFormat = EPG.Utils.GetEpgDateFormat();
            if (string.IsNullOrEmpty(this._EventKeyName))
            {
                ExceptionUtils.ThrowEx(new EAException(ExitCodes.NoValidParameters, "Param EventKeyName Is Empty"));
            }
            else
            {
                try
                {
                    eventToBeModified = EPG.Events[_EventKeyName];
                    EventName = eventToBeModified.Name;
                }
                catch
                {
                    ExceptionUtils.ThrowEx(new EAException(ExitCodes.EventNotExistsFailure, "Event With Key Name - " + this._EventKeyName + " Does Not Exists On Collection"));
                }
            }

            EPGEventdate = eventToBeModified.ConvertedDate;
            Occurences = eventToBeModified.Occurrences;

            if (!string.IsNullOrEmpty(_StartTime) && _StartTime.Contains(":") == false)
            {
                ExceptionUtils.ThrowEx(new EAException(ExitCodes.NoValidParameters, "Start Time Should Be Entered With Format : HH:MM"));
            }

            if (!string.IsNullOrEmpty(_EndTime) && _EndTime.Contains(":") == false)
            {
                ExceptionUtils.ThrowEx(new EAException(ExitCodes.NoValidParameters, "End Time Should Be Entered With Format : HH:MM"));
            }

            Current = false;

            switch (_Frequency)
            {
                case EnumFrequency.ONE_TIME:
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_ONE_TIME");
                    break;
                case EnumFrequency.DAILY:
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_DAILY");
                    break;
                case EnumFrequency.WEEKLY:
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKLY");
                    break;
                case EnumFrequency.WEEKDAY:
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKDAY");
                    break;
                case EnumFrequency.WEEKEND:
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKEND");
                    break;
                case EnumFrequency.NO_CHANGE:
                    Frequency = eventToBeModified.Frequency;
                    break;
                default:
                    ExceptionUtils.ThrowEx(new EAException(ExitCodes.NoValidParameters, "No Valid Frequency Has Been Entered"));
                    break;
            }

            EPG.Live.GetEpgDate(ref EPGDate);
            TodayDateInEPGFormat = DateTime.ParseExact(EPGDate, milestoneDateFormat, CultureInfo.InvariantCulture).ToString(EPGDateFormat);

            if (_Days > 0)
            {
                EPGEventdate = DateTime.ParseExact(EPGEventdate, EPGDateFormat, CultureInfo.InvariantCulture).AddDays(_Days).ToString(EPGDateFormat);
                PCATEventDate = DateTime.ParseExact(EPGEventdate, EPGDateFormat, CultureInfo.InvariantCulture).AddDays(_Days).ToString(PCATDateFormat);
            }

            // Calculate the days delay
            if (EPGEventdate != TodayDateInEPGFormat)
            {
                _Days = (DateTime.ParseExact(EPGEventdate, EPGDateFormat, CultureInfo.InvariantCulture) - DateTime.ParseExact(TodayDateInEPGFormat, EPGDateFormat, CultureInfo.InvariantCulture)).Days;
            }


            EPG.ManualRecording.GetOccurences(Frequency, _Days, ref Occurences);

            if (!string.IsNullOrEmpty(_StartTime) && !string.IsNullOrEmpty(_EndTime))
            {
                newDuration = EPG.Utils._DateTime.SubtractInSec(DateTime.Parse(_EndTime), DateTime.Parse(_StartTime));
            }

            // Tune to service if required
            int channelNum = -1;
            String channelName = "";
            if (!string.IsNullOrEmpty(_ChannelName) && _ChannelName != eventToBeModified.Channel)
            {
                if (tuneToService.ToUpper() == "TRUE")
                {
                    try
                    {
                        channelNum = int.Parse(_manager.GetServiceFromContentXML("Name=" + _ChannelName).LCN);
                    }
                    catch (Exception ex)
                    {
                        ExceptionUtils.ThrowEx(new EAException(ExitCodes.GetChannelFailure, ex.Message));
                    }
                    res = _manager.ChannelSurf(EnumSurfIn.Live, channelNum.ToString());
                    if (!res.CommandSucceeded)
                    {
                        ExceptionUtils.ThrowEx(new EAException(ExitCodes.TuneToChannelFailure, res.FailureReason));
                    }

                }
                else
                {
                    channelName = _ChannelName;
                }

                // Event name should be changed to channel name
                EventName = _ChannelName;

            }
            else
            {
				// For Event based to Time based, fetch channel name from the channel number in event collection
                try
                {
					if (recordNameChangeSupported)
					{
						EventName = _manager.GetServiceFromContentXML("LCN=" + eventToBeModified.Channel).Name;
					}
                }
                catch
                {
                    EPG.Utils.LogCommentInfo("Event channel from collection:" + eventToBeModified.Channel);
                }

            }

            res = _manager.PVR.VerifyEventInPlanner(_EventKeyName);
            if (res.CommandSucceeded == false)
            {
                ExceptionUtils.ThrowEx(new EAException(ExitCodes.FindEventFailure, res.FailureReason));
            }

            EPG.FutureRecordings.SelectEvent(eventToBeModified.Name);

            EPG.FutureRecordings.ModifyEvent();

            //Checking if there are no changes required in setting date and Frequency
            if (EPGEventdate == eventToBeModified.ConvertedDate)
            {
                EPGEventdate = "";
            }

            if (_IsFirstTime)
            {
                if (Frequency == eventToBeModified.Frequency)
                {
                    Frequency = "";
                }
            }

            EPG.ManualRecording.SetManualRecordingParams(EPGEventdate, _StartTime, channelName, channelNum, _EndTime, Frequency, true, _IsFirstTime);

            EPG.ManualRecording.SaveAndEnd(Current, true);

            try
            {
                EPG.ManualRecording.VerifySaveAndEndFinished(Current);
            }
            catch (EAException ex)
            {
                EPG.Events.Remove(_EventKeyName);
                ExceptionUtils.ThrowEx(new EAException(ExitCodes.VerificationFailure, ex.Message));
            }

            EPG.Utils.InsertEventToCollection(this._EventKeyName, EventName, EventSource, _StartTime, _EndTime, _ChannelName, newDuration, 0, PCATEventDate, Frequency,
            Occurences, EPGEventdate, true);

            EPG.Utils.ReturnToLiveViewing();

        }
    }

}

File: UPCH_SAMVGW_INTELGRV_LNUX_02/LittleEndian/IEX/IEX_SRC_FILES/IEX.ElementaryActions.Functionality/EAs/ManualRecording/ModifyManualRecording.vb
Imports FailuresHandler

Namespace EAImplementation
    ''' <summary>
    '''   Modify Manual Recording From Planner
    ''' </summary>
    Public Class ModifyManualRecording
        Inherits IEX.ElementaryActions.BaseCommand

        Private EPG As IEX.ElementaryActions.EPG.SF.UI
        Private _manager As IEX.ElementaryActions.Functionality.Manager

        Private _EventKeyName As String
        Private _StartTime As String
        Private _EndTime As String
        Private _ChannelName As String
        Private _Days As Integer
        Private _Frequency As EnumFrequency

        ''' <param name="EventKeyName">Key Of The Event</param>
        ''' <param name="StartTime">Channel Name</param>
        ''' <param name="EndTime">Channel Number If Entered Doing DCA</param>
        ''' <param name="ChannelName">Optional Parameter Default = "" : The Channel Name</param>
        ''' <param name="Days">Optional Parameter Default = 0 : Adds Days From Current Date</param>
        ''' <param name="Frequency">Optional Parameter Default = ONE_TIME</param>
        ''' <param name="pManager">Manager</param>
        ''' <remarks>
        ''' Possible Error Codes:
        ''' <para>300 - NavigationFailure</para> 
        ''' <para>301 - DictionaryFailure</para> 
        ''' <para>302 - EmptyEPGInfoFailure</para> 
        ''' <para>304 - IRVerificationFailure</para> 
        ''' <para>305 - PCATFailure</para> 
        ''' <para>309 - GetEPGTimeFailure</para>    
        ''' <para>310 - GetEPGDateFailure</para>    
        ''' <para>322 - VerificationFailure</para> 
        ''' <para>328 - INIFailure</para>  
        ''' <para>330 - TelnetFailure</para> 
        ''' <para>331 - CopyFileFailure</para> 
        ''' <para>332 - NoValidParameters</para> 
        ''' <para>334 - VideoNotPresent</para>  
        ''' <para>339 - RecordEventFailure</para>  
        ''' <para>349 - ReturnToLiveFailure</para> 
        ''' <para>350 - ParsingFailure</para> 
        ''' </remarks>
        Sub New(ByVal EventKeyName As String, ByVal StartTime As String, ByVal EndTime As String, ByVal ChannelName As String, _
                ByVal Days As Integer, ByVal Frequency As EnumFrequency, ByVal pManager As IEX.ElementaryActions.Functionality.Manager)

            Me._EventKeyName = EventKeyName
            Me._StartTime = StartTime
            Me._EndTime = EndTime
            Me._ChannelName = ChannelName
            Me._Days = Days
            Me._Frequency = Frequency
            Me._manager = pManager
            EPG = Me._manager.UI
        End Sub

        Protected Overrides Sub Execute()
            If _manager.Project.IsEPGLikeCogeco Then
                ModifyManualFromPlannerCOGECO()
            Else
                ModifyManualFromPlannerUPC()
            End If
        End Sub

        Private Function ModifyManualFromPlannerCOGECO() As Boolean
            Dim res As IEXGateway.IEXResult
            Dim EventDate As String = ""
            Dim EventName As String = ""
            Dim MyStartTime As String = ""
            Dim EndTime As String = ""
            Dim Current As Boolean = True
            Dim EventChannel As String = ""
            Dim EventConvertedDate As String = ""
            Dim EventSource As String = ""
            Dim Frequency As String = ""
            Dim Occurences As Integer = -1

            If Me._EventKeyName = "" Then
                ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "Param EventKeyName Is Empty"))
            Else
                Try
                    EventName = EPG.Events(_EventKeyName).Name
                Catch ex As Exception
                    ExceptionUtils.ThrowEx(New EAException(ExitCodes.EventNotExistsFailure, "Event With Key Name - " + Me._EventKeyName + " Does Not Exists On Collection"))
                End Try
            End If

            If _StartTime <> "" AndAlso _StartTime.Contains(":") = False Then
                ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "Start Time Should Be Entered With Format : HH:MM"))
            End If

            If _EndTime <> "" AndAlso _EndTime.Contains(":") = False Then
                ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "End Time Should Be Entered With Format : HH:MM"))
            End If

            Current = False

            Select Case _Frequency
                Case EnumFrequency.ONE_TIME
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_ONE_TIME")
                    Occurences = 1
                Case EnumFrequency.DAILY
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_DAILY")
                    Occurences = 14
                Case EnumFrequency.WEEKLY
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKLY")
                    Occurences = 2
                Case EnumFrequency.WEEKDAY
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKDAY")
                Case EnumFrequency.WEEKEND
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKEND")
                Case Else
                    ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "No Valid Frequency Has Been Entered"))
            End Select

            EventDate = EPG.Events(_EventKeyName).EventDate

            res = _manager.PVR.VerifyEventInPlanner(_EventKeyName)
            If res.CommandSucceeded = False Then
                ExceptionUtils.ThrowEx(New EAException(res.FailureCode, res.FailureReason))
            End If

            EPG.FutureRecordings.SelectEvent(EventName)

            EPG.FutureRecordings.ModifyEvent()

            Select Case _Frequency
                Case EnumFrequency.ONE_TIME, EnumFrequency.DAILY
                    If _Days > 0 Then
                        Occurences = Occurences - _Days
                    End If
                Case Else
                    If _Days > 7 Then
                        Occurences = 1
                    End If
            End Select

            If _Days <> 0 Then
                Dim ConvertedEpgDate As String = CDate(EventDate).AddDays(_Days).ToString(EPG.Utils.GetEpgDateFormat())
                Dim dateFormatForEventDic As String = EPG.Utils.GetDateFormatForEventDictionary()
                EventDate = CDate(EventDate).AddDays(_Days).ToString(dateFormatForEventDic)

                EPG.ManualRecording.NavigateToDate()

                EPG.ManualRecording.SetDate(ConvertedEpgDate, False)

                EventConvertedDate = ConvertedEpgDate

                EventDate = CDate(EventDate).ToString(dateFormatForEventDic)
            End If

            If _StartTime <> "" Then
                Dim tmpStartTime As Date
                tmpStartTime = CDate(_StartTime)

                EPG.ManualRecording.NavigateToStartTime()

                MyStartTime = tmpStartTime.ToString("HH:mm")

                EPG.ManualRecording.SetHour(tmpStartTime.Hour)

                EPG.ManualRecording.SetMinutes(tmpStartTime.Minute)

                EPG.ManualRecording.VerifyStateId("55")
            End If

            If _EndTime <> "" Then
                Dim tmpEndTime As Date
                tmpEndTime = CDate(_EndTime)

                EPG.ManualRecording.NavigateToEndTime()

                EndTime = tmpEndTime.ToString("HH:mm")

                EPG.ManualRecording.SetHour(tmpEndTime.Hour)

                EPG.ManualRecording.SetMinutes(tmpEndTime.Minute)

                EPG.ManualRecording.VerifyStateId("55")
            End If

            If _Frequency <> EnumFrequency.ONE_TIME Then
                EPG.ManualRecording.NavigateToFrequency()

                EPG.ManualRecording.SetFrequency(Frequency)

                EPG.ManualRecording.VerifyStateId("55")
            End If


            If Me._ChannelName <> "" AndAlso Me._ChannelName <> EPG.Events(_EventKeyName).Channel Then
                EPG.ManualRecording.NavigateToChannel()

                EPG.ManualRecording.SetChannel(Me._ChannelName)

                EPG.ManualRecording.VerifyStateId("55")
            End If

            EPG.ManualRecording.NavigateToRecord(False)

            EventSource = "ManualFuture"


            EPG.ManualRecording.SaveAndEnd(Current)

            Try
                EPG.ManualRecording.VerifySaveAndEndFinished(False)
            Catch ex As EAException
                EPG.Events.Remove(_EventKeyName)
                ExceptionUtils.ThrowEx(New EAException(ex.ExitCode, ex.Message))
            End Try

            EPG.Utils.InsertEventToCollection(Me._EventKeyName, EventName, EventSource, MyStartTime, EndTime, EventChannel, 0, 0, EventDate, Frequency, Occurences, EventConvertedDate, True)

            EPG.Utils.ReturnToLiveViewing()

            Return True
        End Function

        Private Function ModifyManualFromPlannerUPC() As Boolean
            Dim res As IEXGateway.IEXResult
            Dim EventDate As String = ""
            Dim EventName As String = ""
            Dim MyStartTime As String = ""
            Dim EndTime As String = ""
            Dim Current As Boolean = True
            Dim EventChannel As String = ""
            Dim EventConvertedDate As String = ""
            Dim EventSource As String = ""
            Dim Frequency As String = ""
            Dim Occurences As Integer = -1

            If Me._EventKeyName = "" Then
                ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "Param EventKeyName Is Empty"))
            Else
                Try
                    EventName = EPG.Events(_EventKeyName).Name
                Catch ex As Exception
                    ExceptionUtils.ThrowEx(New EAException(ExitCodes.EventNotExistsFailure, "Event With Key Name - " + Me._EventKeyName + " Does Not Exists On Collection"))
                End Try
            End If

            If _StartTime <> "" AndAlso _StartTime.Contains(":") = False Then
                ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "Start Time Should Be Entered With Format : HH:MM"))
            End If

            If _EndTime <> "" AndAlso _EndTime.Contains(":") = False Then
                ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "End Time Should Be Entered With Format : HH:MM"))
            End If

            Current = False

            Select Case _Frequency
                Case EnumFrequency.ONE_TIME
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_ONE_TIME")
                    Occurences = 1
                Case EnumFrequency.DAILY
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_DAILY")
                    Occurences = 14
                Case EnumFrequency.WEEKLY
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKLY")
                    Occurences = 2
                Case EnumFrequency.WEEKDAY
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKDAY")
                Case EnumFrequency.WEEKEND
                    Frequency = EPG.Utils.GetValueFromDictionary("DIC_WEEKEND")
                Case Else
                    ExceptionUtils.ThrowEx(New EAException(ExitCodes.NoValidParameters, "No Valid Frequency Has Been Entered"))
            End Select

            EventDate = EPG.Events(_EventKeyName).EventDate

            res = _manager.PVR.VerifyEventInPlanner(_EventKeyName)
            If res.CommandSucceeded = False Then
                ExceptionUtils.ThrowEx(New EAException(res.FailureCode, res.FailureReason))
            End If

            EPG.FutureRecordings.SelectEvent(EventName)

            EPG.FutureRecordings.ModifyEvent()

            Select Case _Frequency
                Case EnumFrequency.ONE_TIME, EnumFrequency.DAILY
                    If _Days > 0 Then
                        Occurences = Occurences - _Days
                    End If
                Case Else
                    If _Days > 7 Then
                        Occurences = 1
                    End If
            End Select
           

            If _Days > 0 Then
                Dim ConvertedEpgDate As String = CDate(EventDate).AddDays(_Days).ToString(EPG.Utils.GetEpgDateFormat())
                Dim dateFormatForEventDic As String = EPG.Utils.GetDateFormatForEventDictionary()
                EventDate = CDate(EventDate).AddDays(_Days).ToString(dateFormatForEventDic)

                EPG.ManualRecording.NavigateToDate()

                EPG.ManualRecording.SetDate(ConvertedEpgDate, False)

                EventConvertedDate = ConvertedEpgDate

                EventDate = CDate(EventDate).ToString(dateFormatForEventDic)
            End If

            If _StartTime <> "" Then
                Dim tmpStartTime As Date
                tmpStartTime = CDate(_StartTime)

                EPG.ManualRecording.NavigateToStartTime()

                MyStartTime = tmpStartTime.ToString("HH:mm")

                EPG.ManualRecording.SetStartTime(tmpStartTime.ToString("HHmm"))

                EPG.ManualRecording.VerifyStateId("52")
            End If

            If _EndTime <> "" Then
                Dim tmpEndTime As Date
                tmpEndTime = CDate(_EndTime)

                EPG.ManualRecording.NavigateToEndTime()

                EndTime = tmpEndTime.ToString("HH:mm")

                EPG.ManualRecording.SetEndTime(tmpEndTime.ToString("HHmm"))
            End If

            If _Frequency <> EnumFrequency.ONE_TIME Then
                EPG.ManualRecording.NavigateToFrequency()

                EPG.ManualRecording.SetFrequency(Frequency)
            End If


            If Me._ChannelName <> "" AndAlso Me._ChannelName <> EPG.Events(_EventKeyName).Channel Then
                EPG.ManualRecording.NavigateToChannel()

                EPG.ManualRecording.SetChannel(Me._ChannelName)
            End If

            EPG.ManualRecording.NavigateToRecord(True)

            EventSource = "ManualFuture"

            EPG.ManualRecording.SaveAndEnd(Current)

            Try
                EPG.ManualRecording.VerifySaveAndEndFinished(False)
            Catch ex As EAException
                EPG.Events.Remove(_EventKeyName)
                ExceptionUtils.ThrowEx(New EAException(ex.ExitCode, ex.Message))
            End Try

            EPG.Utils.InsertEventToCollection(Me._EventKeyName, EventName, EventSource, MyStartTime, EndTime, EventChannel, 0, 0, EventDate, Frequency, Occurences, EventConvertedDate, True)

            EPG.Utils.ReturnToLiveViewing()

            Return True
        End Function
    End Class


End Namespace